src/backend/distributed/planner/distributed_planner.c

```diff
@@ -97,12 +98,49 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 	Query *originalQuery = NULL;
 	PlannerRestrictionContext *plannerRestrictionContext = NULL;
 	bool setPartitionedTablesInherited = false;
+	bool constraintCheckQuery = false;
 
 	if (cursorOptions & CURSOR_OPT_FORCE_DISTRIBUTED)
 	{
 		needsDistributedPlanning = true;
 	}
 
+	if (AlterTableInProgress() && parse->commandType == CMD_SELECT &&
+		needsDistributedPlanning)
+	{
+		constraintCheckQuery = true;
+
+		/*
+		 * Disable distributed planning for ALTER TABLE constraint validation
+		 * queries. These constraints will be validated in worker nodes, so
+		 * no need to run expensive distributed queries from coordinator.
+		 *
+		 * For example, ALTER TABLE ... ATTACH PARTITION checks that the new
+		 * partition doesn't violate constraints of the parent table, which
+		 * might involve running some SELECT queries. Ideally we'd completely
+		 * skip these checks in the coordinator, but we don't have any means to
+		 * tell postgres to skip the checks. So the best we can do is to execute
+		 * these queries locally.
+		 */
+		needsDistributedPlanning = false;
```

Comment:
do we need both this and the LIMIT 0 or would LIMIT 0 be sufficient? (could we simplify the code?)

Reply:
This is sufficient, but we would get `push down of limit count: 0` debug messages after `ALTER TABLE`s, which will be confusing.

I am trying to see if we can do this in `ExecutorRun()` as you suggested.